# Attack 1 - DNS Flood: Comparison

Loads baseline and attack1 CSV exports and compares them (volume, rates, deltas).
See `attacks/attack1-dns-flood.md` for the attack description and deployment manifests.

- Baseline capture: `exports/baseline/`
- Attack capture: `exports/attack1-dns-flood/`
- Comparison logic: `compare_logs.py`

In [2]:
from IPython.display import display

%run compare_logs.py

print("=== Resume ===")
display(summary)

print("\n=== Events per second (stats) ===")
display(eps_stats)

print("\n=== Resource x Verb ===")
display(resource_x_verb)

for field, df in results.items():
    print(f"\n=== {field} ===")
    display(df)

        métrique  baseline  attack  variation
0   total_events      2089    2449        360
1  patch_on_pods         0       0          0
=== Resume ===


,métrique,baseline,attack,variation
0,total_events,2089,2449,360
1,patch_on_pods,0,0,0



=== Events per second (stats) ===


,métrique,baseline,attack,variation
0,max_events_par_sec,25.0,30.00,5.00
1,moyenne_events_par_sec,7.6,9.57,1.97



=== Resource x Verb ===


,resource,verb,count_baseline,count_attack,variation
0,pods,get,0,180,180
1,nodes,get,87,174,87
2,configmaps,watch,21,36,15
3,events,patch,0,11,11
4,clusterroles,watch,0,6,6
...,...,...,...,...,...
67,ippools,watch,6,0,-6
68,pods,watch,12,6,-6
69,services,watch,24,18,-6
70,validatingadmissionpolicies,watch,6,0,-6



=== verb ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,get,1063,1302,212.6,260.4,239,47.8
1,update,710,704,142.0,140.8,-6,-1.2
2,watch,309,303,61.8,60.6,-6,-1.2
3,patch,3,17,0.6,3.4,14,2.8
4,create,0,4,0.0,0.8,4,0.8
5,list,4,4,0.8,0.8,0,0.0



=== objectRef_resource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,leases,713,707,142.6,141.4,-6,-1.2
1,nodes,108,204,21.6,40.8,96,19.2
2,pods,12,186,2.4,37.2,174,34.8
3,endpointslices,70,72,14.0,14.4,2,0.4
4,endpoints,64,66,12.8,13.2,2,0.4
...,...,...,...,...,...,...,...
58,ipreservations,3,0,0.6,0.0,-3,-0.6
59,ippools,6,0,1.2,0.0,-6,-1.2
60,kubecontrollersconfigurations,3,0,0.6,0.0,-3,-0.6
61,validatingadmissionpolicies,6,0,1.2,0.0,-6,-1.2



=== objectRef_namespace ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,kube-system,643,658,128.6,131.6,15,3.0
1,safe-ns,0,183,0.0,36.6,183,36.6
2,default,116,120,23.2,24.0,4,0.8
3,kube-node-lease,88,87,17.6,17.4,-1,-0.2
4,attacker-ns,0,3,0.0,0.6,3,0.6



=== objectRef_subresource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,exec,0,106,0.0,21.2,106,21.2
1,status,3,6,0.6,1.2,3,0.6



=== responseStatus_code ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,200,990,1097,198.0,219.4,107,21.4
1,101,0,69,0.0,13.8,69,13.8



=== authorization_decision ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,allow,990,1166,198.0,233.2,176,35.2



=== authorization_reason ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,"RBAC: allowed by ClusterRoleBinding ""system:pu...",330,314,66.0,62.8,-16,-3.2
1,"RBAC: allowed by ClusterRoleBinding ""system:ku...",236,237,47.2,47.4,1,0.2
2,"RBAC: allowed by ClusterRoleBinding ""system:ku...",161,162,32.2,32.4,1,0.2
3,"RBAC: allowed by ClusterRoleBinding ""kubeadm:c...",0,106,0.0,21.2,106,21.2
4,"RBAC: allowed by ClusterRoleBinding ""system:di...",65,69,13.0,13.8,4,0.8
5,"RBAC: allowed by ClusterRoleBinding ""calico-ku...",30,34,6.0,6.8,4,0.8
6,"RBAC: allowed by ClusterRoleBinding ""kubeadm:n...",14,14,2.8,2.8,0,0.0
7,"RBAC: allowed by ClusterRoleBinding ""system:vo...",5,3,1.0,0.6,-2,-0.4
8,"RBAC: allowed by RoleBinding ""system::extensio...",0,2,0.0,0.4,2,0.4
9,"RBAC: allowed by ClusterRoleBinding ""system:pu...",5,1,1.0,0.2,-4,-0.8



=== sourceIPs ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,10.10.10.10,1572,1768,314.4,353.6,196,39.2
1,::1,210,210,42.0,42.0,0,0.0
2,192.168.62.159,0,126,0.0,25.2,126,25.2
3,10.10.10.11,83,116,16.6,23.2,33,6.6
4,10.10.10.12,89,114,17.8,22.8,25,5.0
5,10.0.2.15,6,0,1.2,0.0,-6,-1.2
6,192.168.62.160,9,0,1.8,0.0,-9,-1.8
7,192.168.62.161,120,0,24.0,0.0,-120,-24.0



=== user_groups ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:authenticated,1429,1706,285.8,341.2,277,55.4
1,system:unauthenticated,660,628,132.0,125.6,-32,-6.4
2,system:nodes,217,336,43.4,67.2,119,23.8
3,system:serviceaccounts,251,227,50.2,45.4,-24,-4.8
4,system:serviceaccounts:kube-system,251,227,50.2,45.4,-24,-4.8
5,system:masters,210,210,42.0,42.0,0,0.0
6,kubeadm:cluster-admins,0,180,0.0,36.0,180,36.0



=== user_username ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:anonymous,660,628,132.0,125.6,-32,-6.4
1,system:kube-controller-manager,428,429,85.6,85.8,1,0.2
2,system:kube-scheduler,323,324,64.6,64.8,1,0.2
3,system:apiserver,210,210,42.0,42.0,0,0.0
4,kubernetes-admin,0,180,0.0,36.0,180,36.0
5,system:serviceaccount:kube-system:calico-kube-...,120,126,24.0,25.2,6,1.2
6,system:node:k8s-cp,72,121,14.4,24.2,49,9.8
7,system:node:k8s-w1,71,110,14.2,22.0,39,7.8
8,system:node:k8s-w2,74,105,14.8,21.0,31,6.2
9,system:serviceaccount:kube-system:generic-garb...,40,40,8.0,8.0,0,0.0



=== userAgent ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,kube-probe/1.30,660,628,132.0,125.6,-32,-6.4
1,kubelet/v1.30.14 (linux/amd64) kubernetes/9e18483,217,336,43.4,67.2,119,23.8
2,kube-controller-manager/v1.30.14 (linux/amd64)...,296,294,59.2,58.8,-2,-0.4
3,kube-scheduler/v1.30.14 (linux/amd64) kubernet...,296,294,59.2,58.8,-2,-0.4
4,kube-apiserver/v1.30.14 (linux/amd64) kubernet...,210,210,42.0,42.0,0,0.0
5,kubectl/v1.30.14 (linux/amd64) kubernetes/9e18483,0,180,0.0,36.0,180,36.0
6,kube-controller-manager/v1.30.14 (linux/amd64)...,96,99,19.2,19.8,3,0.6
7,kube-controllers/v0.0.0 (linux/amd64) kubernet...,60,66,12.0,13.2,6,1.2
8,Go-http-client/2.0,72,60,14.4,12.0,-12,-2.4
9,kube-controller-manager/v1.30.14 (linux/amd64)...,40,40,8.0,8.0,0,0.0


## Results summary

| Metric | Baseline | Attack | Variation | Factor |
|---|---|---|---|---|
| Total audit events | 2089 | 2449 | +360 | x1.17 |
| Events/sec (average) | 7.6 | 9.57 | +1.97 | x1.26 |
| Events/sec (peak) | 25 | 30 | +5 | x1.2 |
| `patch` on `pods` | 0 | 0 | 0 | - |

| Verb | Baseline | Attack | Delta |
|---|---|---|---|
| get | 1063 | 1302 | +239 |
| update | 710 | 704 | -6 |
| watch | 309 | 303 | -6 |
| patch | 3 | 17 | +14 |
| create | 0 | 4 | +4 |

| New entity (absent from baseline) | Count in attack |
|---|---|
| `userAgent: kubectl/v1.30.14` | 180 |
| `user.username: kubernetes-admin` | 180 |
| `user.groups: kubeadm:cluster-admins` | 180 |
| `objectRef.subresource: exec` | 106 |
| `responseStatus.code: 101` | 69 |

| Namespace | Baseline | Attack |
|---|---|---|
| safe-ns | 0 | 183 |
| attacker-ns | 0 | 3 |
| kube-system | 643 | 658 |

---

## Analysis of results

The overall volume barely moves: total events grow by only 17% (2089 to 2449), and two of the five verbs, `update` and `watch`, actually decrease slightly during the attack. This is consistent with the attack mechanism itself: the flood is a raw UDP socket flood against CoreDNS's DNS service (port 53), sent directly from the attacker pods without going through the `kube-apiserver` at any point, so the malicious traffic itself leaves no trace in these logs by construction.

Nearly all of the modest increase that does appear traces back to a single source that has nothing to do with the attack: `kubectl`, `kubernetes-admin`, and `kubeadm:cluster-admins` each jump from 0 to 180 occurrences, together with `objectRef.subresource: exec` (0 to 106) and `responseStatus.code: 101` (0 to 69, the HTTP code for a protocol upgrade, consistent with an interactive `kubectl exec` session). These four fields describe the same thing from different angles: a `kubectl exec` session run as `kubernetes-admin`, which matches exactly how the measurement pipeline works, `capture-metrics-loop.sh` polls CoreDNS's metrics endpoint by executing into the victim pod. The `pods get` count (0 to 180 in the resource/verb breakdown) lines up with this too, since `kubectl exec` first fetches the target pod object before opening the exec session. `objectRef.namespace: safe-ns` rising from 0 to 183 confirms the same thing from the namespace side, since the victim pod that the measurement script execs into lives in `safe-ns`.

Taken together, these fields account for the large majority of the observed delta on their own. `patch` (+14) and `create` (+4) are small enough in absolute terms to plausibly be ordinary baseline variation rather than attack-driven. `patch_on_pods` stays at 0 in both scenarios, confirming no pod status activity is associated with this attack.

No field in this dataset carries a signal that originates from the DNS flood itself. The attacker pods, their UDP traffic, and CoreDNS's internal rejection mechanism (`MaxConcurrentRejectCount`) are entirely absent from the Kubernetes audit trail, which only sees API server activity.

---

## Detectability analysis

**This attack is not detectable in Kubernetes audit logs.** Every indicator that moves during the attack capture window traces back to the measurement methodology itself (`kubectl exec` polling through the victim pod), not to the attacker's traffic.

**The near-zero native signal is itself informative for this study.** The contrast between a 17% rise in total audit events and the severe impact measured through other telemetry (CPU, packet loss, `nf_conntrack` saturation) demonstrates concretely that Kubernetes audit logging has a blind spot for attacks that operate below the API server, at the kernel/network level.

### Detectability limitations to note

- **Measurement artifacts can masquerade as signal.** The appearance of `kubectl`, `kubernetes-admin`, and `pods/exec` in this capture is a direct result of how the attack's impact was measured (`kubectl exec`-based polling), not of the attack. Any pipeline that captures audit logs alongside its own measurement tooling needs to account for this before attributing observed deltas to the attack under test.
- **No substitute indicator is available in this log source.** Unlike attacks that manipulate Kubernetes objects directly, this attack produces no usable `objectRef`, `verb`, or `userAgent` signal at all. Detecting it requires a different telemetry source entirely, such as CoreDNS's own metrics (`MaxConcurrentRejectCount`) or `nf_conntrack` occupancy, both already used to measure this attack's impact in this study.